### PR TITLE
SAMZA-2522: Add metadata generator for azure storage blob

### PR DESCRIPTION
**Pull request by lakshmi-manasa-g**
**Feature:** This PR extends the existing AzureBlobSystemProducer by adding the ability to enhance an azure storage blob with metadata - basically key-value pairs to the blob's metadata (sits outside of the blob's actual content). This aids tremendously in ingesting these blobs into Kusto (Azure's data explorer). 

**Changes:** New interfaces BlobMetadataGeneratorFactory, BlobMetadataGenerator and class BlobMetadataContext added where the user needs to implement BlobMetadataGeneratorFactory and BlobMetadataGenerator which takes in a BlobMetadataContext containing the stream name and the size of the current blob being committed. An instance of the generator will be created when the blob is about to be committed. Generator is invoked to get metadata and the returned metadata is attached to the blob.
Any additional configs (key:value pairs) needed for this generator can be passed as with prefix systems.<system-name>.azureblob.metadataGeneratorConfig.\<key\> with value \<value\>. 

A default impl is provided which does not add anything to the blob.

**API changes:** New interfaces BlobMetadataGeneratorFactory, BlobMetadataGenerator and class BlobMetadataContext added. New configs with prefix systems.<system-name>.azureblob.metadataGeneratorConfig. The factory is wired in through systems.<system-name>.azureblob.metadataPropertiesGeneratorFactory. 
Prior to this PR, "rawSizeBytes" metadata was always added  by default to the blobs committed. Now, that will not be the case. A BlobMetadataGenerator implementation has to be provided to add this. 

**Upgrade instructions:** Backwards incompatible. "rawSizeBytes" metadata will no longer be added by default. Instead a generator will have to be provided. 

**Usage instructions:** Wire in the generator factory through systems.<system-name>.azureblob.metadataPropertiesGeneratorFactory and pass in additional configs with prefix systems.<system-name>.azureblob.metadataGeneratorConfig 

**Tests:** Existing tests updated to assert for metadata attached.

### CODE REVIEW

File: samza-azure/src/main/java/org/apache/samza/system/azureblob/AzureBlobConfig.java [hunk 1]
```diff
@@ -94,6 +95,18 @@
...
+  // this factory should return an implementation of org.apache.samza.system.azureblob.utils.BlobMetadataGenerator
+  // this generator will be invoked when a blob is committed to add metadata properties to it
+  public static final String SYSTEM_BLOB_METADATA_PROPERTIES_GENERATOR_FACTORY = SYSTEM_AZUREBLOB_PREFIX + "metadataPropertiesGeneratorFactory";
+  private static final String SYSTEM_BLOB_METADATA_PROPERTIES_GENERATOR_FACTORY_DEFAULT =
+      "org.apache.samza.system.azureblob.utils.NullBlobMetadataGeneratorFactory";
+
+  // Additional configs for the metadata generator should be prefixed with this string which is passed to the generator.
+  // for example, to pass a "key":"value" pair to the metadata generator, add config like
+  // systems.<system-name>.azureblob.metadataGeneratorConfig.<key> with value <value>
+  public static final String SYSTEM_BLOB_METADATA_GENERATOR_CONFIG_PREFIX = SYSTEM_AZUREBLOB_PREFIX + "metadataGeneratorConfig";
```
Comment (cameronlee314) [inline samza-azure/src/main/java/org/apache/samza/system/azureblob/AzureBlobConfig.java:108]: Does this need to have an extra `.` at the end so it gets stripped off in `getSystemMetadataGeneratorConfigs`?

File: samza-azure/src/main/java/org/apache/samza/system/azureblob/AzureBlobConfig.java [hunk 2]
```diff
@@ -185,4 +198,13 @@ public long getMaxBlobSize(String systemName) {
...
     return getLong(String.format(SYSTEM_MAX_MESSAGES_PER_BLOB, systemName), SYSTEM_MAX_MESSAGES_PER_BLOB_DEFAULT);
   }
+
+  public String getSystemMetadataPropertiesGeneratorFactory(String systemName) {
+    return get(String.format(SYSTEM_BLOB_METADATA_PROPERTIES_GENERATOR_FACTORY, systemName),
+        SYSTEM_BLOB_METADATA_PROPERTIES_GENERATOR_FACTORY_DEFAULT);
+  }
+
+  public Config getSystemMetadataGeneratorConfigs(String systemName) {
+    return subset(String.format(SYSTEM_BLOB_METADATA_GENERATOR_CONFIG_PREFIX, systemName));
```
Comment (cameronlee314) [inline samza-azure/src/main/java/org/apache/samza/system/azureblob/AzureBlobConfig.java:208]: Minor: For consistency, consider naming these `getSystemBlobMetadata...`

File: samza-azure/src/main/java/org/apache/samza/system/azureblob/avro/AzureBlobOutputStream.java [hunk 3]
```diff
@@ -172,8 +193,8 @@ public synchronized void close() {
       future.get((long) flushTimeoutMs, TimeUnit.MILLISECONDS);
       LOG.info("For blob: {} committing blockList size:{}", blobAsyncClient.getBlobUrl().toString(), blockList.size());
       metrics.updateAzureCommitMetrics();
-      Map<String, String> blobMetadata = Collections.singletonMap(BLOB_RAW_SIZE_BYTES_METADATA, Long.toString(totalUploadedBlockSize));
```
Comment (cameronlee314) [inline samza-azure/src/main/java/org/apache/samza/system/azureblob/avro/AzureBlobOutputStream.java:175]: If an existing case needs this BLOB_RAW_SIZE_BYTES_METADATA, then will they need to implement their own blob metadata generator?

Comment (lakshmi-manasa-g) [reply samza-azure/src/main/java/org/apache/samza/system/azureblob/avro/AzureBlobOutputStream.java:175]: ah, yes. they will need to implement their own generator. Updated the PR description to reflect backwards incompatible.  is that ok?

**Comment by lakshmi-manasa-g** [comment]
the exception is happening in Samza-core_2.11:test whereas these changes are all in samza-azure module. Hence these changes can not be causing the build failure seen in https://travis-ci.org/github/apache/samza/builds/687272039. Also triggered another build with an empty commit.

**Pull request merged**